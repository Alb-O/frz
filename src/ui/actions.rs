use anyhow::Result;
use ratatui::crossterm::event::{KeyCode, KeyEvent};

use crate::types::{SearchMode, SearchOutcome};

use super::App;

impl<'a> App<'a> {
    pub(crate) fn handle_key(&mut self, key: KeyEvent) -> Result<Option<SearchOutcome>> {
        match key.code {
            KeyCode::Esc => {
                return Ok(Some(SearchOutcome {
                    accepted: false,
                    selection: None,
                    query: self.search_input.text().to_string(),
                }));
            }
            KeyCode::Enter => {
                let selection = self.current_selection();
                return Ok(Some(SearchOutcome {
                    accepted: true,
                    selection,
                    query: self.search_input.text().to_string(),
                }));
            }
            KeyCode::Tab => {
                self.mark_query_dirty();
                self.switch_mode();
            }
            KeyCode::Up => {
                self.move_selection_up();
            }
            KeyCode::Down => {
                self.move_selection_down();
            }
            _ => {
                if self.search_input.input(key) {
                    self.mark_query_dirty();
                    self.request_search();
                }
            }
        }
        Ok(None)
    }

    fn switch_mode(&mut self) {
        self.mode = match self.mode {
            SearchMode::Facets => SearchMode::Files,
            SearchMode::Files => SearchMode::Facets,
        };
        self.table_state.select(Some(0));
        self.request_search();
    }

    fn move_selection_up(&mut self) {
        if let Some(selected) = self.table_state.selected()
            && selected > 0
        {
            self.table_state.select(Some(selected - 1));
        }
    }

    fn move_selection_down(&mut self) {
        if let Some(selected) = self.table_state.selected() {
            let len = self.filtered_len();
            if selected + 1 < len {
                self.table_state.select(Some(selected + 1));
            }
        }
    }
}
