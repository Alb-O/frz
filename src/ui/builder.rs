#[cfg(feature = "fs")]
use std::sync::mpsc::Receiver;

use std::collections::HashMap;

use anyhow::Result;
#[cfg(not(feature = "fs"))]
use anyhow::bail;
use ratatui::layout::Constraint;

use super::App;
use crate::plugins::SearchPluginRegistry;
#[cfg(feature = "fs")]
use crate::systems::filesystem::{FilesystemOptions, IndexUpdate, spawn_filesystem_index};
use crate::theme::Theme;
use crate::types::{SearchData, SearchMode, SearchOutcome, UiConfig};

/// A small builder for configuring the interactive search UI.
/// This presents an fzf-like API for setting prompts, column
/// headings and column widths before running the interactive picker.
pub struct SearchUi {
    data: SearchData,
    input_title: Option<String>,
    headers: HashMap<SearchMode, Vec<String>>,
    widths: HashMap<SearchMode, Vec<Constraint>>,
    ui_config: Option<UiConfig>,
    theme: Option<Theme>,
    start_mode: Option<SearchMode>,
    plugins: SearchPluginRegistry,
    #[cfg(feature = "fs")]
    index_updates: Option<Receiver<IndexUpdate>>,
}

impl SearchUi {
    /// Create a new search UI for the provided data.
    pub fn new(data: SearchData) -> Self {
        Self {
            data,
            input_title: None,
            headers: HashMap::new(),
            widths: HashMap::new(),
            ui_config: None,
            theme: None,
            start_mode: None,
            plugins: SearchPluginRegistry::default(),
            #[cfg(feature = "fs")]
            index_updates: None,
        }
    }

    /// Create a search UI pre-populated with files from the filesystem rooted at `path`.
    #[cfg(feature = "fs")]
    pub fn filesystem(path: impl AsRef<std::path::Path>) -> Result<Self> {
        Self::filesystem_with_options(path.as_ref().to_path_buf(), FilesystemOptions::default())
    }

    #[cfg(feature = "fs")]
    pub fn filesystem_with_options(
        path: impl Into<std::path::PathBuf>,
        options: FilesystemOptions,
    ) -> Result<Self> {
        let root = path.into();
        let (data, updates) = spawn_filesystem_index(root, options)?;
        let mut ui = Self::new(data);
        ui.start_mode = Some(SearchMode::FILES);
        ui.index_updates = Some(updates);
        Ok(ui)
    }

    #[cfg(not(feature = "fs"))]
    pub fn filesystem(_path: impl AsRef<std::path::Path>) -> Result<Self> {
        bail!("filesystem support is disabled; enable the `fs` feature");
    }

    pub fn with_input_title(mut self, title: impl Into<String>) -> Self {
        self.input_title = Some(title.into());
        self
    }

    pub fn with_headers_for(self, mode: SearchMode, headers: Vec<&str>) -> Self {
        let headers = headers.into_iter().map(|s| s.to_string()).collect();
        self.with_headers(mode, headers)
    }

    pub fn with_widths_for(mut self, mode: SearchMode, widths: Vec<Constraint>) -> Self {
        self.widths.insert(mode, widths);
        self
    }

    pub fn with_ui_config(mut self, config: UiConfig) -> Self {
        self.ui_config = Some(config);
        self
    }

    pub fn with_initial_query(mut self, query: impl Into<String>) -> Self {
        self.data.initial_query = query.into();
        self
    }

    pub fn with_theme_name(mut self, name: &str) -> Self {
        if let Some(theme) = crate::theme::by_name(name) {
            self.theme = Some(theme);
        }
        self
    }

    pub fn with_theme(mut self, theme: Theme) -> Self {
        self.theme = Some(theme);
        self
    }

    pub fn with_start_mode(mut self, mode: SearchMode) -> Self {
        self.start_mode = Some(mode);
        self
    }

    /// Replace the plugin registry driving the search worker.
    pub fn with_plugin_registry(mut self, plugins: SearchPluginRegistry) -> Self {
        self.plugins = plugins;
        self
    }

    /// Mutably configure the plugin registry.
    pub fn configure_plugins<F>(mut self, configure: F) -> Self
    where
        F: FnOnce(&mut SearchPluginRegistry),
    {
        configure(&mut self.plugins);
        self
    }

    /// Run the interactive search UI with the configured options.
    #[cfg_attr(not(feature = "fs"), allow(unused_mut))]
    pub fn run(mut self) -> Result<SearchOutcome> {
        // Build an App and apply optional customizations, then run it.
        let mut app = App::with_plugins(self.data, self.plugins);
        if let Some(title) = self.input_title {
            app.input_title = Some(title);
        }
        for (mode, headers) in self.headers {
            app.set_headers_for(mode, headers);
        }
        for (mode, widths) in self.widths {
            app.set_widths_for(mode, widths);
        }
        if let Some(ui) = self.ui_config {
            app.ui = ui;
            app.ensure_tab_buffers();
        }
        if let Some(theme) = self.theme {
            app.set_theme(theme);
        }
        if let Some(mode) = self.start_mode {
            app.set_mode(mode);
        }
        #[cfg(feature = "fs")]
        if let Some(updates) = self.index_updates.take() {
            app.set_index_updates(updates);
        }

        app.run()
    }

    fn with_headers(mut self, mode: SearchMode, headers: Vec<String>) -> Self {
        self.headers.insert(mode, headers);
        self
    }
}
