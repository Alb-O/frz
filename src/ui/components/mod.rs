//! UI building blocks shared across rendering and state modules.

pub mod progress;
pub mod tables;
pub mod tabs;

pub use progress::IndexProgress;
pub use tables::{TablePane, render_table};
pub use tabs::{InputContext, ProgressState, render_input_with_tabs};
