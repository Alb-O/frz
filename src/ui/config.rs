use std::collections::HashMap;

use frz_plugin_api::{SearchMode, descriptors::SearchPluginDescriptor};

/// Labels and titles for one of the search panes.
#[derive(Debug, Clone)]
pub struct PaneUiConfig {
    pub mode_title: String,
    pub hint: String,
    pub table_title: String,
    pub count_label: String,
}

impl PaneUiConfig {
    /// Construct a new [`PaneUiConfig`] with custom labels.
    #[must_use]
    pub fn new(
        mode_title: impl Into<String>,
        hint: impl Into<String>,
        table_title: impl Into<String>,
        count_label: impl Into<String>,
    ) -> Self {
        Self {
            mode_title: mode_title.into(),
            hint: hint.into(),
            table_title: table_title.into(),
            count_label: count_label.into(),
        }
    }
}

/// Full UI definition for a contributed tab.
#[derive(Debug, Clone)]
pub struct TabUiConfig {
    pub mode: SearchMode,
    pub tab_label: String,
    pub pane: PaneUiConfig,
}

impl TabUiConfig {
    /// Build a [`TabUiConfig`] from constituent pieces.
    #[must_use]
    pub fn new(mode: SearchMode, tab_label: impl Into<String>, pane: PaneUiConfig) -> Self {
        Self {
            mode,
            tab_label: tab_label.into(),
            pane,
        }
    }
}

/// Text used by the UI when rendering pane content and tab labels.
#[derive(Debug, Clone)]
pub struct UiConfig {
    pub filter_label: String,
    pub detail_panel_title: String,
    tabs: Vec<TabUiConfig>,
    index: HashMap<SearchMode, usize>,
}

impl Default for UiConfig {
    fn default() -> Self {
        let mut config = Self {
            filter_label: "Filter facets".to_string(),
            detail_panel_title: "Selection details".to_string(),
            tabs: Vec::new(),
            index: HashMap::new(),
        };
        for descriptor in crate::plugins::builtin::descriptors() {
            config.register_plugin(descriptor);
        }
        config
    }
}

impl UiConfig {
    /// UI configuration tailored to searching tags and files.
    #[must_use]
    pub fn tags_and_files() -> Self {
        let mut config = Self {
            filter_label: "Filter tag".to_string(),
            detail_panel_title: "Selection details".to_string(),
            tabs: Vec::new(),
            index: HashMap::new(),
        };

        let facets = crate::plugins::builtin::facets::mode();
        config.register_tab(TabUiConfig::new(
            facets,
            "Tags",
            PaneUiConfig::new(
                "Tag search",
                "Type to filter tags. Press Tab to view files.",
                "Matching tags",
                "Tags",
            ),
        ));

        let files = crate::plugins::builtin::files::mode();
        config.register_tab(TabUiConfig::new(
            files,
            "Files",
            PaneUiConfig::new(
                "File search",
                "Type to filter files by path or tag. Press Tab to view tags.",
                "Matching files",
                "Files",
            ),
        ));

        config
    }

    /// Register a new tab definition with this configuration.
    pub fn register_tab(&mut self, tab: TabUiConfig) {
        let mode = tab.mode;
        if let Some(position) = self.index.get(&mode).copied() {
            self.tabs[position] = tab;
        } else {
            let idx = self.tabs.len();
            self.index.insert(mode, idx);
            self.tabs.push(tab);
        }
    }

    /// Register the default UI provided by a plugin descriptor.
    pub fn register_plugin(&mut self, descriptor: &'static SearchPluginDescriptor) {
        let mode = SearchMode::from_descriptor(descriptor);
        let pane = PaneUiConfig::new(
            descriptor.ui.mode_title,
            descriptor.ui.hint,
            descriptor.ui.table_title,
            descriptor.ui.count_label,
        );
        self.register_tab(TabUiConfig::new(mode, descriptor.ui.tab_label, pane));
    }

    /// Return all registered tabs in the order they were added.
    #[must_use]
    pub fn tabs(&self) -> &[TabUiConfig] {
        &self.tabs
    }

    /// Lookup tab metadata for the provided mode.
    #[must_use]
    pub fn tab(&self, mode: SearchMode) -> Option<&TabUiConfig> {
        self.index
            .get(&mode)
            .and_then(|position| self.tabs.get(*position))
    }

    /// Lookup pane metadata for the provided mode.
    #[must_use]
    pub fn pane(&self, mode: SearchMode) -> Option<&PaneUiConfig> {
        self.tab(mode).map(|tab| &tab.pane)
    }

    /// Mutably lookup pane metadata for the provided mode.
    pub fn pane_mut(&mut self, mode: SearchMode) -> Option<&mut PaneUiConfig> {
        let position = self.index.get(&mode).copied()?;
        self.tabs.get_mut(position).map(|tab| &mut tab.pane)
    }

    /// Retrieve the label displayed on the tab itself.
    #[must_use]
    pub fn tab_label(&self, mode: SearchMode) -> Option<&str> {
        self.tab(mode).map(|tab| tab.tab_label.as_str())
    }

    /// Resolve a registered mode identifier to its [`SearchMode`].
    #[must_use]
    pub fn mode_by_id(&self, id: &str) -> Option<SearchMode> {
        self.tabs
            .iter()
            .find(|tab| tab.mode.id() == id)
            .map(|tab| tab.mode)
    }

    /// Return the pane title associated with the provided mode.
    #[must_use]
    pub fn mode_title(&self, mode: SearchMode) -> &str {
        self.pane(mode)
            .map(|pane| pane.mode_title.as_str())
            .unwrap_or("")
    }

    /// Return the hint text associated with the provided mode.
    #[must_use]
    pub fn mode_hint(&self, mode: SearchMode) -> &str {
        self.pane(mode).map(|pane| pane.hint.as_str()).unwrap_or("")
    }

    /// Return the table title associated with the provided mode.
    #[must_use]
    pub fn mode_table_title(&self, mode: SearchMode) -> &str {
        self.pane(mode)
            .map(|pane| pane.table_title.as_str())
            .unwrap_or("")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::plugins::builtin::{facets, files};

    #[test]
    fn tags_and_files_registers_tabs() {
        let config = UiConfig::tags_and_files();
        assert!(config.tab(facets::mode()).is_some());
        assert!(config.tab(files::mode()).is_some());
    }
}
