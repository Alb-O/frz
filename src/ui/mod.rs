//! Interactive terminal UI components and helpers.
//!
//! `builder` exposes the public-facing [`SearchUi`] builder, while the other
//! submodules implement the event loop, rendering, and state management that
//! power the terminal application.

mod actions;
mod builder;
mod config;
mod indexing;
mod render;
mod runtime;
mod search;
mod state;

pub use builder::SearchUi;
pub use config::{PaneUiConfig, TabUiConfig, UiConfig};
pub use runtime::run;
pub use state::App;
