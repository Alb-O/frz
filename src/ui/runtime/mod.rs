mod event_loop;
mod hydration;

pub use event_loop::run;

#[cfg(test)]
mod tests;
