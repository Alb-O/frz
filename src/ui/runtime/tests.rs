use crate::plugins::api::{AttributeRow, FileRow, SearchData, search::Fs};
use crate::plugins::builtin::files;
use crate::systems::filesystem::{IndexUpdate, ProgressSnapshot};
use crate::ui::App;
use ratatui::{Terminal, backend::TestBackend};
use std::io;
use std::path::{Path, PathBuf};
use std::sync::mpsc;

fn sample_index_update() -> IndexUpdate {
    IndexUpdate {
        files: vec![
            FileRow::filesystem("src/lib.rs", ["src", "*.rs"]),
            FileRow::filesystem("src/main.rs", ["src", "*.rs"]),
            FileRow::filesystem("README.md", ["*.md"]),
        ]
        .into(),
        attributes: vec![
            AttributeRow::new("*.md", 1),
            AttributeRow::new("*.rs", 2),
            AttributeRow::new("src", 2),
        ]
        .into(),
        progress: ProgressSnapshot {
            indexed_attributes: 3,
            indexed_files: 3,
            total_attributes: Some(3),
            total_files: Some(3),
            complete: true,
        },
        reset: true,
        cached_data: None,
    }
}

#[test]
fn initial_files_tab_render_captures_missing_results() {
    let mut app = App::new(SearchData::new());
    app.set_mode(files::mode());
    app.hydrate_initial_results();

    let (tx, rx) = mpsc::channel();
    app.set_index_updates(rx);
    tx.send(sample_index_update()).unwrap();

    app.pump_index_updates();
    app.pump_search_results();

    let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
    terminal.draw(|frame| app.draw(frame)).unwrap();

    let view = {
        let backend = terminal.backend();
        backend.to_string()
    };
    insta::assert_snapshot!("initial_files_tab_render_captures_missing_results", view);

    assert!(
        app.filtered_len() > 0,
        "expected initial search results to populate without any user input"
    );
}

struct MassiveSyntheticFs {
    total: usize,
}

impl MassiveSyntheticFs {
    fn new(total: usize) -> Self {
        Self { total }
    }
}

struct MassiveIter {
    remaining: usize,
    index: usize,
}

impl Iterator for MassiveIter {
    type Item = io::Result<PathBuf>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.index >= self.remaining {
            return None;
        }

        let dir = self.index / 512;
        let path = PathBuf::from(format!("dir_{dir:04}/file_{:06}.txt", self.index));
        self.index += 1;
        Some(Ok(path))
    }
}

impl Fs for MassiveSyntheticFs {
    type Iter = MassiveIter;

    fn walk(&self, _root: &Path) -> io::Result<Self::Iter> {
        Ok(MassiveIter {
            remaining: self.total,
            index: 0,
        })
    }
}

#[test]
fn massive_filesystem_initial_load_shows_preview_snapshot() {
    const TOTAL_FILES: usize = 125_000;

    let fs = MassiveSyntheticFs::new(TOTAL_FILES);
    let data = SearchData::from_filesystem_with(&fs, Path::new("/synthetic")).unwrap();
    assert!(
        data.files.len() >= 100_000,
        "expected synthetic filesystem to exceed 100k entries"
    );
    let total_files = data.files.len();
    let total_attributes = data.attributes.len();
    const PREVIEW_SLICE: usize = 512;
    let preview_files: Vec<FileRow> = data.files.iter().take(PREVIEW_SLICE).cloned().collect();
    let preview_attributes = data.attributes.clone();
    drop(data);

    let mut app = App::new(SearchData::new());
    app.set_mode(files::mode());
    app.hydrate_initial_results();

    assert_eq!(
        app.filtered_len(),
        0,
        "no results should be visible before indexing begins"
    );

    let (tx, rx) = mpsc::channel();
    app.set_index_updates(rx);
    tx.send(IndexUpdate {
        files: preview_files.into(),
        attributes: preview_attributes.into(),
        progress: ProgressSnapshot {
            indexed_attributes: total_attributes,
            indexed_files: PREVIEW_SLICE,
            total_attributes: Some(total_attributes),
            total_files: Some(total_files),
            complete: false,
        },
        reset: true,
        cached_data: None,
    })
    .unwrap();

    app.pump_index_updates();
    app.pump_search_results();
    assert!(
        app.filtered_len() > 0,
        "expected preview results to be visible during indexing"
    );

    app.throbber_state.calc_next();

    let mut terminal = Terminal::new(TestBackend::new(100, 30)).unwrap();
    terminal.draw(|frame| app.draw(frame)).unwrap();

    let view = {
        let backend = terminal.backend();
        backend.to_string()
    };

    insta::assert_snapshot!(
        "massive_filesystem_initial_load_shows_preview_snapshot",
        view
    );
}
