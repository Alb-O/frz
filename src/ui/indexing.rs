use std::sync::mpsc::{Receiver, TryRecvError};
use std::thread;
use std::time::{Duration, Instant};

// Indexing work intentionally runs under strict per-tick limits so UI rendering stays
// responsive even when large trees are being ingested. `MAX_INDEX_UPDATES_PER_TICK`
// bounds how many incremental updates we merge in a single frame, while
// `MAX_INDEX_PROCESSING_TIME` caps the wall-clock time spent applying updates before we
// yield back to drawing and input handling.

use crate::systems::filesystem::{IndexUpdate, merge_update};
use frz_plugin_api::SearchData;

use super::App;
use frz_tui::components::IndexProgress;

impl<'a> App<'a> {
    const MAX_INDEX_UPDATES_PER_TICK: usize = 32;
    const MAX_INDEX_PROCESSING_TIME: Duration = Duration::from_millis(8);

    pub(crate) fn set_index_updates(&mut self, updates: Receiver<IndexUpdate>) {
        self.index_updates = Some(updates);
        if self.data.attributes.is_empty() && self.data.files.is_empty() {
            self.index_progress = IndexProgress::with_unknown_totals();
        } else {
            self.index_progress
                .refresh_from_data(&self.data, self.dataset_totals());
        }
    }

    pub(crate) fn pump_index_updates(&mut self) {
        let Some(rx) = self.index_updates.take() else {
            return;
        };

        let mut should_request = false;
        let mut keep_receiver = true;
        let mut processed = 0usize;
        let start = Instant::now();

        loop {
            if processed >= Self::MAX_INDEX_UPDATES_PER_TICK
                || start.elapsed() >= Self::MAX_INDEX_PROCESSING_TIME
            {
                break;
            }
            match rx.try_recv() {
                Ok(mut update) => {
                    let cached_data = update.cached_data.take();
                    self.notify_search_of_update(&update);
                    should_request |= self.apply_index_update(update, cached_data);
                    processed += 1;
                }
                Err(TryRecvError::Empty) => break,
                Err(TryRecvError::Disconnected) => {
                    keep_receiver = false;
                    break;
                }
            }
        }

        if keep_receiver {
            self.index_updates = Some(rx);
        }

        if should_request {
            let waiting_for_initial = self.filtered_len() == 0;
            if waiting_for_initial {
                self.initial_results_deadline = Some(Instant::now() + Duration::from_millis(250));
            }
            self.request_search_after_index_update();
            if waiting_for_initial {
                self.wait_for_initial_results();
            }
        }
    }

    fn apply_index_update(&mut self, update: IndexUpdate, cached_data: Option<SearchData>) -> bool {
        let mut changed = false;

        match cached_data {
            Some(data) => {
                self.data = data;
                for state in self.tab_states.values_mut() {
                    state.filtered.clear();
                    state.scores.clear();
                }
                self.table_state.select(None);
                self.index_progress
                    .refresh_from_data(&self.data, self.dataset_totals());
                self.mark_query_dirty();
                changed = true;
            }
            None => {
                if update.reset {
                    self.index_progress = IndexProgress::with_unknown_totals();
                    for state in self.tab_states.values_mut() {
                        state.filtered.clear();
                        state.scores.clear();
                    }
                    self.table_state.select(None);
                }

                let update_changed =
                    update.reset || !update.files.is_empty() || !update.attributes.is_empty();
                if update_changed {
                    merge_update(&mut self.data, &update);
                    self.mark_query_dirty();
                    changed = true;
                }
            }
        }

        let progress = update.progress;
        let attributes_key = crate::plugins::builtin::attributes::descriptor().id;
        let files_key = crate::plugins::builtin::files::descriptor().id;
        self.index_progress.record_indexed(&[
            (attributes_key, progress.indexed_attributes),
            (files_key, progress.indexed_files),
        ]);
        self.index_progress.set_totals(&[
            (attributes_key, progress.total_attributes),
            (files_key, progress.total_files),
        ]);
        if progress.complete {
            self.index_progress.mark_complete();
        }

        changed
    }

    fn wait_for_initial_results(&mut self) {
        let Some(deadline) = self.initial_results_deadline else {
            return;
        };

        if Instant::now() >= deadline {
            self.initial_results_deadline = None;
            return;
        }

        while Instant::now() < deadline {
            self.pump_search_results();
            if self.filtered_len() > 0 {
                return;
            }

            if !self.search.is_in_flight() {
                thread::sleep(Duration::from_millis(10));
                continue;
            }

            thread::sleep(Duration::from_millis(10));
        }

        self.pump_search_results();
        if Instant::now() >= deadline {
            self.initial_results_deadline = None;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::systems::filesystem::ProgressSnapshot;
    use frz_plugin_api::{AttributeRow, FileRow, SearchData};
    use std::time::{Duration, Instant};

    fn wait_for_results(app: &mut App) {
        let deadline = Instant::now() + Duration::from_secs(1);
        while app.search.is_in_flight() && Instant::now() < deadline {
            std::thread::sleep(Duration::from_millis(10));
            app.pump_search_results();
        }
        app.pump_search_results();
    }

    #[test]
    fn index_updates_refresh_results_without_input_changes() {
        let data = SearchData::new();
        let mut app = App::new(data);

        app.mark_query_dirty();
        app.request_search();
        wait_for_results(&mut app);

        assert_eq!(app.filtered_len(), 0);
        let update = IndexUpdate {
            files: vec![FileRow::filesystem("src/lib.rs", ["alpha"])].into(),
            attributes: vec![AttributeRow::new("alpha", 1)].into(),
            progress: ProgressSnapshot {
                indexed_attributes: 1,
                indexed_files: 1,
                total_attributes: Some(1),
                total_files: Some(1),
                complete: true,
            },
            reset: false,
            cached_data: None,
        };

        app.notify_search_of_update(&update);
        let changed = app.apply_index_update(update, None);
        assert!(changed, "index update should report data changes");
        assert!(
            app.search.has_unapplied_input(),
            "data changes should mark the query dirty"
        );

        app.request_search_after_index_update();
        wait_for_results(&mut app);

        assert_eq!(
            app.filtered_len(),
            0,
            "results should remain stable until the user edits the query"
        );

        app.mark_query_dirty_from_user_input();
        app.request_search();
        wait_for_results(&mut app);

        assert!(
            app.filtered_len() > 0,
            "expected refreshed results after user input"
        );
    }
}
