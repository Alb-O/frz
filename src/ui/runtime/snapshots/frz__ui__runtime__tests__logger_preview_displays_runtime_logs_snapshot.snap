---
source: src/ui/runtime/tests.rs
assertion_line: 230
expression: view
---
" Log viewer > Indexed attributes: 0 • Indexed Files: 0 • Indexed Lo  Tags   Files   Logs  "
"                                      Logger controls                                     "
" ┌Targets──────────────────┐┌Runtime log [log=<rate>]───────────────────────────────────┐ "
" │EWIDT:frz::preview::image││<time>:D:frz::preview::image:src/ui/runtime/tests.rs:<line>:│ "
" │EWIDT:frz::tests::logger ││         image transport negotiated sixel protocol         │ "
" │                         ││<time>:D:frz::preview::image:src/ui/runtime/tests.rs:<line>:│ "
" │                         ││         image transport enabled debug diagnostics         │ "
" │                         ││<time>:E:frz::tests::logger:src/ui/runtime/tests.rs:<line>:f│ "
" │                         ││         ailed to attach to remote debugger: connection ref│ "
" │                         ││         used                                              │ "
" │                         ││<time>:W:frz::tests::logger:src/ui/runtime/tests.rs:<line>:f│ "
" │                         ││         alling back to offline analysis mode              │ "
" │                         ││<time>:I:frz::tests::logger:src/ui/runtime/tests.rs:<line>:q│ "
" │                         ││         ueued 3 indexing tasks for the diagnostics subsyst│ "
" │                         ││         em                                                │ "
" │                         ││<time>:D:frz::tests::logger:src/ui/runtime/tests.rs:<line>:p│ "
" │                         ││         oll cycle completed with 3 queued tasks           │ "
" │                         ││<time>:I:frz::preview::image:src/ui/runtime/tests.rs:<line>:│ "
" │                         ││         tui graphics backend initialised for snapshot test│ "
" │                         ││         ing                                               │ "
" │                         ││<time>:D:frz::preview::image:src/ui/runtime/tests.rs:<line>:│ "
" │                         ││         kitty graphics handshake acknowledged in debug mod│ "
" │                         ││         e                                                 │ "
" │                         ││                                                           │ "
" └─────────────────────────┘└───────────────────────────────────────────────────────────┘ "
